base-server.d.ts

```diff
@@ -0,0 +1,704 @@
+   */
+  id?: string
+
+  /**
+   * Development or production server mode. By default,
+   * it will be taken from `NODE_ENV` environment variable.
+   * On empty `NODE_ENV` it will be `'development'`.
+   */
+  env?: 'production' | 'development'
+
+  /**
+   * Process ID, to display in reporter.
+   */
+  pid?: number
+
+  /**
+   * HTTP server to connect WebSocket server to it. Same as in `ws.Server`.
+   */
+  server?: HTTPServer
+
+  /**
+   * Port to bind server. It will create HTTP server manually to connect
+   * WebSocket server to it. Default is `31337`.
+   */
+  port?: number
+
+  /**
+   * IP-address to bind server. Default is `127.0.0.1`.
+   */
+  host?: string
+
+  /**
+   * SSL key or path to it. Path could be relative from server root.
+   * It is required in production mode, because WSS is highly recommended.
+   */
+  key?: string
+
+  /**
+   * SSL certificate or path to it. Path could be relative from server
+   * root. It is required in production mode, because WSS is highly
+   * recommended.
+   */
+  cert?: string
+
+  /**
+   * Function to show current server status.
+   */
+  reporter?: (event: string, payload: Object) => void
+}
+
+/**
+ * Basic Logux Server API without good UI. Use it only if you need
+ * to create some special hacks on top of Logux Server.
+ *
+ * In most use cases you should use {@link Server}.
+ *
+ * ```js
+ * const { BaseServer } = require('@logux/server')
+ * class MyLoguxHack extends BaseServer {
+ *   …
+ * }
+ * ```
+ */
+export class BaseServer {
+  constructor(opts: LoguxBaseServerOptions)
+
+  /**
+   * Server options.
+   *
+   * ```js
+   * console.log('Server options', server.options.subprotocol)
+   * ```
+   */
+  options: LoguxBaseServerOptions
+  reporter: LoguxBaseServerOptions['reporter'] | (() => void)
+
+  /**
+   * Production or development mode.
+   *
+   * ```js
+   * if (server.env === 'development') {
+   *   logDebugData()
+   * }
+   * ```
+   */
+  env: Required<LoguxBaseServerOptions['env']>
+
+  /**
+   * Server unique ID.
+   *
+   * ```js
+   * console.log('Error was raised on ' + server.nodeId)
+   * ```
+   */
+  nodeId: string
+
+  /**
+   * Server actions log.
+   *
+   * ```js
+   * server.log.each(finder)
+   * ```
+   */
+  log: Log
+
+  /**
+   * Connected clients.
+   *
+   * ```js
+   * for (let i in server.connected) {
+   *   console.log(server.connected[i].remoteAddress)
+   * }
+   * ```
+   */
+  connected: {
+    [key: string]: ServerClient
+  }
+
+  /**
+   * Set authenticate function. It will receive client credentials
+   * and node ID. It should return a Promise with `true` or `false`.
+   *
+   * @param authenticator The authentication callback.
+   *
+   * ```js
+   * server.auth(async (userId, token) => {
+   *   const user = await findUserByToken(token)
+   *   return !!user && userId === user.id
+   * })
+   * ```
+   */
+  auth<Credentials = any>(authenticator: LoguxAuthenticator<Credentials>): void
```

Comment:
Let’s limit `credentials` to string as I did in Logux Client 
https://github.com/logux/client/blob/master/client/index.d.ts#L51